presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -643,20 +644,29 @@ public PlanWithProperties visitStatisticsWriterNode(StatisticsWriterNode node, P
         @Override
         public PlanWithProperties visitTableFinish(TableFinishNode node, PreferredProperties preferredProperties)
         {
-            PlanWithProperties child = planChild(node, PreferredProperties.any());
+            PlanNode child = planChild(node, PreferredProperties.any()).getNode();
 
-            // if the child is already a gathering exchange, don't add another
-            if ((child.getNode() instanceof ExchangeNode) && ((ExchangeNode) child.getNode()).getType().equals(GATHER)) {
-                return rebaseAndDeriveProperties(node, child);
+            ExchangeNode gather;
+            // in case the input is a union (see PushTableWriteThroughUnion)
+            if (child instanceof ExchangeNode) {
+                ExchangeNode exchangeNode = (ExchangeNode) child;
```

Comment:
Note after the change you don't check whether the type is `GATHER`

Reply:
Also, when reverting (https://github.com/prestodb/presto/pull/12526) in https://github.com/prestodb/presto/pull/12622, I have observed some weird
 query shape in corner cases: 

![73224273_1148337928710854_2940914628995055616_n](https://user-images.githubusercontent.com/799346/67826307-4f55d700-fa89-11e9-96dc-02fb2cba9e46.jpg)

Can such case be handled correctly? 

Since such plan shape is logically correct, we should either be able to handle it, or having check state to guard from such shape (which will be difficult to find all corner cases I guess)

Reply:
We should always expect the exchange before the `TableFinishOperator` to be gather. If there's any other type of exchange - it feels like it can always be replaced with gather. Theres no reason of doing `Exchange[PARTITION] -> Exchange[GATHER]` right before the `TableFinishOperator`.

Reply:
As discussed, since recoverable grouped execution would only be enabled when the parent fragment contains `TableFinishOperator`, so fragment 3 in this example won't be running in recoverable mode, and thus the sequence for this fragment is not a must. It is not a blocker from recoverable grouped execution and page ordering perspective.

Reply:
@arhimondr : Shall, but this is a legit plan. So either and state check (which will fail certain queries), or have some way to handle this. 

But as @shixuan-fan mentioned, it's not a blocker for recoverable grouped execution but some "better engineering" fix :)